src/ncp/ncp_base.cpp

```diff
@@ -3494,13 +3544,53 @@ ThreadError NcpBase::SetPropertyHandler_MAC_SCAN_STATE(uint8_t header, spinel_pr
             break;
 
         case SPINEL_SCAN_STATE_ENERGY:
-            errorCode = otEnergyScan(
-                mInstance,
-                mChannelMask,
-                mScanPeriod,
-                &HandleEnergyScanResult_Jump,
-                this
-            );
+#if OPENTHREAD_ENABLE_RAW_LINK_API
+            if (otLinkRawIsEnabled(mInstance))
+            {
+                if (mCurScanChannel == NCP_INVALID_SCAN_CHANNEL)
+                {
+                    uint8_t scanChannel;
+                    uint16_t scanDuration;
+
+                    parsedLength = spinel_datatype_unpack(
```

Comment:
Sorry for not chiming in often on stuff like this recently, I've been slammed with design work on an entirely different system.

My biggest problem with this change is that you are changing the behavior of a well established mechanism for what amounts to an implementation detail. The implementation should achieve the design, not the other way around. Your implementation might only scan one channel at a time, but that is not a limitation that Spinel puts on the energy scan feature you are using. But keep reading.

A while back when we were talking about this I voiced the thought that the energy scan behavior should live in the host, which seems to be what you are trying to enable here. The difference was that I was imagining setting the channel and then repeatedly sampling RSSI, much like how an MCU might try to do an energy scan on a dumb transceiver IC.

However, in that case you can poll RSSI and get a response in a single SPI transaction, making it fast. With Spinel, you aren't going to have that sort of quick turn-around, so I admit that particular idea was not the greatest. There is a need for some sort of NCP-side helper for energy scans. I can also see why the existing energy scan mechanism might seem a bit clunky under those circumstances.

Ultimately, if you are going to use the existing energy scan mechanism, you need to adhere to the established behavior of that mechanism. This means that you would perform a scan on all of the given channels and return the results asynchronously, finishing by changing the property `PROP_MAC_SCAN_STATE` to `SCAN_STATE_IDLE`.

BUT... I think likely the best option would be to define a new property called `PROP_PHY_MAX_RSSI` (open to suggestions), which uses `PROP_PHY_CHAN` for the channel and `PROP_MAC_SCAN_PERIOD` for the duration. You query this property and you get back a result with the same transaction ID, making it much more convenient to use.

Alternatively, another option would be to fail if there was more than one channel present in the channel mask. But you'd still need to emit `PROP_MAC_SCAN_STATE` as `SCAN_STATE_IDLE` after you are done (which you aren't doing here).

-------------

Below are some other thoughts from earlier drafts of replying to this issue, included because I figured there might be some useful thoughts in there somewhere:

> It seemed inefficient to separate out a single function call into multiple individual commands just to pass the arguments. The Host is then required to make several async function calls (to send commands) and wait for their results in series.

Spinel is designed with the following philosophy:

* If you can reasonably implement a behavior with multiple properties, prefer that over implementing new commands. (There are cases where using properties would be a lot more convoluted than just using commands, such as the [host buffer offload](https://rawgit.com/openthread/openthread/master/doc/draft-spinel-protocol.html#rfc.section.9) feature))
* Avoid using structures for properties which aren't lists---such structures should instead be broken into multiple properties. 

These are intentional parts of Spinel's design which help ensure backward compatibility. Otherwise we shouldn't be using properties, we should be using commands.

Granted, what is "convoluted" and what is "reasonable" can be a matter of opinion, in which case we can discuss the matter as we are doing here.

> Besides scans, I couldn't find another OpenThread API that split it's calling logic between multiple function calls. For instance, all the _STREAM commands take all parameters inline.

The `*_STREAM` *properties* (they are not commands) are structures because we don't have any other way to ensure all of the individual attributes are associated together as one atomic unit.